### DISCUSSION TITLE
loadAstNode

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-langium/langium#1747] [Q&A] [answered]
the code in `buildReference` where the element is `isAstNodeDescription`

calls `loadAstNode` which calls ` this.langiumDocuments().getDocument`

what is the proper way to make this work with lazy documents?

**Comment by msujew** [comment]
Hey @cdietrich,

if the document isn't in memory yet, you would need to load it from disk and parse it. Likely via this method:

https://github.com/eclipse-langium/langium/blob/21b105164186dfff6a4763c21d450931e2200e5b/packages/langium/src/workspace/documents.ts#L138

Since this can happen in the process of resolving a variable (which is happening in sync contexts), loading the file needs to happen in sync as well.

**Reply by cdietrich** [reply]
so i need to override getDocument?

**Reply by cdietrich** [reply]
e.g. 
```
export class SampleLangiumDocuments extends DefaultLangiumDocuments {

    override getDocument(uri: URI): LangiumDocument | undefined {
        const result = super.getDocument(uri);
        if (result === undefined) {
           const content = fs.readFileSync(uri.fsPath, 'utf-8')
           const  document = this.langiumDocumentFactory.fromString(content, uri)
           this.addDocument(document)
           return document;
        }
        return result;
    }

}
```

**Reply by cdietrich** [reply]
what i dont get though it goes into the if also if the document is present. need to check why

**Reply by cdietrich** [reply]
getOrCreateDocument calls getDocument ....

**Reply by msujew** [reply]
You should override `loadAstNode`. Loading a document dynamically is currently not supported directly in the framework.

**Reply by cdietrich** [reply, marked as answer]
ok

```
export class SampleLinker extends DefaultLinker {
    protected override loadAstNode(nodeDescription: AstNodeDescription): AstNode | undefined {
        if (nodeDescription.node) {
            return nodeDescription.node;
        }
        let doc = this.langiumDocuments().getDocument(nodeDescription.documentUri);
        if (!doc) {
            const content = fs.readFileSync(nodeDescription.documentUri.fsPath, 'utf-8')
            doc = this.langiumDocuments().createDocument(nodeDescription.documentUri, content);
            if (!doc) {
                return undefined;
            }
           
        }
        return this.astNodeLocator.getAstNode(doc.parseResult.value, nodeDescription.path);
    }
}
```
seems to work as well.

**Reply by msujew** [reply]
Cool :)

**Reply by spoenemann** [reply]
Thought: shall we add an _optional_ method to synchronously read files in the FileSystemProvider? Then we could add something like the above as default behavior and it would work at least in a Node.js context.
